### PR TITLE
#15 deactivate outdated items

### PR DESCRIPTION
**Pull request by axce1**
add two commands:
- alert_before_remove_item
- remove_old_items

### CODE REVIEW

File: dcf/management/commands/alert_before_remove_item.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+            - timezone.timedelta(days=settings.DCF_ALERT_BEFORE_REMOVE_DAYS)
+
+        old_items = Item.objects.values('id', 'user_id')\
+            .annotate(last_change=Max('updated'))\
+            .filter(last_change__lt=remove_before, is_active=1)\
+            .order_by()
+
+        # group items by user
+        userdict = defaultdict(list)
+        for _ in old_items:
```
Comment (inoks) [inline dcf/management/commands/alert_before_remove_item.py]: bad variable naming

File: dcf/management/commands/alert_before_remove_item.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+            userdict[_['user_id']].append(_['id'])
+
+        for user in userdict:
```
Comment (inoks) [inline dcf/management/commands/alert_before_remove_item.py]: Can you just iterate over QuerySet of Users, which have outdated items?

File: proj/settings.py [hunk 3]
```diff
@@ -223,6 +226,13 @@
     'PAGE_SIZE': 10
 }
 
+EMAIL_HOST = 'smtp.example.com'
```
Comment (inoks) [inline proj/settings.py]: Should keep `settings` clean and not add any `EMAIL_` settings here.

File: dcf/management/commands/remove_old_items.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+
+    def handle(self, *args, **options):
+
+        remove_before = timezone.now() \
+            - timezone.timedelta(days=settings.DCF_ITEM_MAX_AGE_DAYS)
+
+        old_items = Item.objects.values('id', 'user_id')\
+           .annotate(last_change=Max('updated'))\
+           .filter(last_change__lte=remove_before.date(), is_active=1)\
```
Comment (inoks) [inline dcf/management/commands/remove_old_items.py]: Should use compare with full datetime value, not only date()
`is_active=1` - Should compare with boolean value.

File: dcf/management/commands/remove_old_items.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+           .order_by()
+
+        users = CustomUser.objects.filter\
+            (id__in=old_items.values('user_id'))
+
+        for user in users:
```
Comment (inoks) [inline dcf/management/commands/remove_old_items.py]: Probably better have to query:
1) Get all affected users
2) Get all affected items per user
I think that would be more clear.

File: dcf/tests/test_commands.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        management.call_command('loaddata', 'dcf/tests/init_data_for_command_test.json', verbosity=0)
+
+
+class TestCommands(GenericViewTestCase):
+
+    def test_alert(self):
+
+        mail.outbox=[]
+
+        out = StringIO()
```
Comment (inoks) [inline dcf/tests/test_commands.py]: I think we need to check:
1) Outdated item link are in mail body
2) Non outdated item link not in mail body

File: dcf/tests/test_commands.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+        management.call_command('alert_before_remove_item')
+        self.assertEqual(len(mail.outbox), 3)
+        self.assertEquals(out.getvalue(), '')
+
+    def test_remove(self):
+
+        out = StringIO()
+        management.call_command('remove_old_items')
+        self.assertEquals(out.getvalue(), '')
+        items = Item.objects.filter(is_active=0).count()
```
Comment (inoks) [inline dcf/tests/test_commands.py]: I think we need to check:

Outdated item deactivated
Non outdated item not deactivated
Outdated item link are in mail body
Non outdated item link not in mail body

File: dcf/management/commands/alert_before_remove_item.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+        for user in users_obj:
+            items_url_list = []
+
+            # get items objects
+            items_obj = Item.objects.filter\
+                (id__in=old_items.values('id'),
+                 user_id=user.id)
+
+            for item in items_obj:
+                full_item_url = ''.join([settings.DCF_SITE_SCHEMA,
```
Comment (inoks) [inline dcf/management/commands/alert_before_remove_item.py]: use `format()` for string construction

File: dcf/templates/email_remove.html [hunk 9]
```diff
@@ -0,0 +1,8 @@
+{% load i18n %}
+
+{% trans "Your items:" %}
```
Comment (inoks) [inline dcf/templates/email_remove.html]: `.po` translation file should be update too.

File: dcf/tests/init_data_for_command_test.json [hunk 10]
```diff
@@ -0,0 +1 @@
...0", "phone": "8999999999", "is_active": true, "updated": "2017-04-28T03:06:45.462Z", "posted": "2017-05-03T03:06:45.462Z"}}, {"model": "dcf.item", "pk": 2, "fields": {"slug": "item_2", "user": 1, "group": 1, "title": "Item #2", "description": "Item #2", "price": "1212.00", "phone": "8999999999", "is_active": true, "updated": "2017-05-03T03:07:21.797Z", "posted": "2017-05-03T03:07:21.797Z"}}, {"model": "dcf.item", "pk": 3, "fields": {"slug": "item_3", "user": 1, "group": 1, "title": "Item #3", "description": "Item #3", "price": "1212.00", "phone": "8999999999", "is_active": true, "updated": "2017-04-03T03:07:21.797Z", "posted": "2017-04-03T03:07:21.797Z"}}, {"model": "dcf.item", "pk": 4, "fields": {"slug": "item_4", "user": 2, "group": 1, "title": "Item #4", "description": "Item #4", "price": "111.00", "phone": "8999999999", "is_active": true, "updated": "2017-04-28T03:06:45.462Z", "posted": "2017-05-03T03:06:45.462Z"}}, {"model": "dcf.item", "pk": 5, "fields": {"slug": "item_5", "user": 3, "group": 1, "title": "Item #5", "description": "Item #5", "price": "10.00", "phone": "822222222", "is_active": true, "updated": "2016-06-28T03:06:45.462Z", "posted": "2016-05-03T03:06:45.462Z"}}]
```
Comment (inoks) [inline dcf/tests/init_data_for_command_test.json]: Better create test data at TestCase  setUp()

Comment (axce1) [inline dcf/tests/init_data_for_command_test.json]: yep, but i cannt change updated field item
or i not found solution yet

File: dcf/management/commands/remove_old_items.py (hunk 5 above)
Comment (axce1) [inline dcf/management/commands/remove_old_items.py]: > Get all affected items per user
i get all items per user in loop 
wrong?

File: dcf/templates/email_remove.html [hunk 11]
```diff
@@ -0,0 +1,8 @@
...
+{% for item in items %}
+  {{ item }}
+{% endfor %}
+{% trans "have been deleted." %}
```
Comment (axce1) [inline dcf/templates/email_remove.html]: s/deleted/deactivated  ?

Comment (inoks) [inline dcf/templates/email_remove.html]: Items was marked as inactive, and was not deleted.

File: dcf/tests/init_data_for_command_test.json (hunk 10 above)
Comment (inoks) [inline dcf/tests/init_data_for_command_test.json]: You can do that either creating inactive items or just mark some test items as inactive after you created them.

File: dcf/management/commands/remove_old_items.py (hunk 5 above)
Comment (inoks) [inline dcf/management/commands/remove_old_items.py]: I don't think the way you do that is clear enough.
Why don't you use smth like that:
```python
users = CustomUser.objects.filter(item__updated__lte=_your_date_), item__is_active=True)
```
?

Comment (axce1) [inline dcf/management/commands/remove_old_items.py]: good

File: dcf/tests/init_data_for_command_test.json (hunk 10 above)
Comment (axce1) [inline dcf/tests/init_data_for_command_test.json]: > You can do that either creating inactive items or just mark some test items as inactive after you created them.

cant create inactive or mark items inactive, query check is_active=True

**Pull request closed without merging**